std-bits/table/src/main/java/org/enso/table/data/column/operation/map/numeric/arithmetic/NumericBinaryOpImplementation.java

```diff
@@ -200,160 +202,156 @@ protected DoubleStorage runDoubleMap(
+          builder.appendLong(r);
         }
       }
 
       context.safepoint();
     }
 
-    if (m < n) {
-      isNothing.set(m, n);
-    }
-
-    return new LongStorage(out, n, isNothing, INTEGER_RESULT_TYPE);
+    return builder.seal();
   }
 
   protected Storage<Long> runLongMap(
       AbstractLongStorage a, Long b, MapOperationProblemAggregator problemAggregator) {
     if (b == null) {
-      return LongStorage.makeEmpty(a.size(), INTEGER_RESULT_TYPE);
+      return LongStorage.makeEmpty(a.getSize(), INTEGER_RESULT_TYPE);
     }
 
     long bNonNull = b;
     Context context = Context.getCurrent();
-    int n = a.size();
-    long[] out = new long[n];
-    BitSet isNothing = new BitSet();
-    for (int i = 0; i < n; i++) {
+    long n = a.getSize();
+    var builder = Builder.getForLong(INTEGER_RESULT_TYPE, n, problemAggregator);
+    for (long i = 0; i < n; i++) {
       if (a.isNothing(i)) {
-        isNothing.set(i);
+        builder.appendNulls(1);
       } else {
-        Long r = doLong(a.getItem(i), bNonNull, i, problemAggregator);
+        Long r = doLong(a.getItemAsLong(i), bNonNull, i, problemAggregator);
         if (r == null) {
-          isNothing.set(i);
+          builder.appendNulls(1);
         } else {
-          out[i] = r;
+          builder.appendLong(r);
         }
       }
 
       context.safepoint();
     }
 
-    return new LongStorage(out, n, isNothing, INTEGER_RESULT_TYPE);
+    return builder.seal();
   }
 
-  protected BigIntegerStorage runBigIntegerZip(
+  protected Storage<BigInteger> runBigIntegerZip(
       BigIntegerArrayAdapter a,
       BigIntegerArrayAdapter b,
       MapOperationProblemAggregator problemAggregator) {
     Context context = Context.getCurrent();
-    int n = a.size();
-    int m = Math.min(a.size(), b.size());
-    BigInteger[] out = new BigInteger[n];
-    for (int i = 0; i < m; i++) {
+    long n = a.size();
+    long m = Math.min(n, b.size());
+    var builder = Builder.getForBigInteger(n, BlackholeProblemAggregator.INSTANCE);
+    for (long i = 0; i < m; i++) {
       BigInteger x = a.getItem(i);
       BigInteger y = b.getItem(i);
       if (x != null && y != null) {
-        BigInteger r = doBigInteger(x, y, i, problemAggregator);
-        out[i] = r;
+        builder.append(doBigInteger(x, y, i, problemAggregator));
       }
       context.safepoint();
     }
 
-    return new BigIntegerStorage(out, n);
+    if (m < n) {
+      builder.appendNulls(Math.toIntExact(n - m));
+    }
+
+    return builder.seal();
   }
 
-  protected BigIntegerStorage runBigIntegerMap(
+  protected Storage<BigInteger> runBigIntegerMap(
       BigIntegerArrayAdapter a, BigInteger b, MapOperationProblemAggregator problemAggregator) {
     Context context = Context.getCurrent();
-    int n = a.size();
-    BigInteger[] out = new BigInteger[n];
-    for (int i = 0; i < n; i++) {
+    long n = a.size();
+    var builder = Builder.getForBigInteger(n, BlackholeProblemAggregator.INSTANCE);
```

Comment:
Same as above.

Even if now it doesn't affect the actual behaviour, I think we should be passed the `problemAggregator` wherever it is required.

Otherwise - in the future we can change the BigIntegerBuilder to actually start using it more (because it can, since it already holds a reference to the aggregator). Then we will start running into bugs where the problems are somehow not aggregated in some random places like here.

I think the default should be - whenever a method takes a `problemAggregator`, we SHOULD pass to it an actual aggregator, unless we have a very good reason to pass the `Blackhole` one.

Reply:
Agree missed that I had one here.